src/Features/Core/Diagnostics/AbstractHostDiagnosticUpdateSource.cs

```diff
@@ -0,0 +1,133 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract class AbstractHostDiagnosticUpdateSource : IDiagnosticUpdateSource
+    {
+        private static ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>> _analyzerHostDiagnosticsMap =
+            ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>>.Empty;
+
+        protected AbstractHostDiagnosticUpdateSource()
+        {
+            // Register for exception diagnostics from workspace's analyzer manager.
+            WorkspaceAnalyzerManager.AnalyzerExceptionDiagnostic += OnAnalyzerExceptionDiagnostic;
+        }
+
+        ~AbstractHostDiagnosticUpdateSource()
+        {
+            // Unregister for exception diagnostics from workspace's analyzer manager.
+            WorkspaceAnalyzerManager.AnalyzerExceptionDiagnostic -= OnAnalyzerExceptionDiagnostic;
+        }
+
+        protected abstract Workspace Workspace { get; }
+
+        public bool SupportGetDiagnostics
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        public ImmutableArray<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId projectId, DocumentId documentId, object id, CancellationToken cancellationToken)
+        {
+            return ImmutableArray<DiagnosticData>.Empty;
+        }
+
+        public event EventHandler<DiagnosticsUpdatedArgs> DiagnosticsUpdated;
+
+        protected void RaiseDiagnosticsUpdated(DiagnosticsUpdatedArgs args)
+        {
+            var updated = this.DiagnosticsUpdated;
+            if (updated != null)
+            {
+                updated(this, args);
+            }
+        }
+
+        private void OnAnalyzerExceptionDiagnostic(object sender, WorkspaceAnalyzerExceptionDiagnosticArgs args)
+        {
+            if (this.Workspace != args.Workspace)
+            {
+                return;
+            }
+
+            Contract.ThrowIfFalse(AnalyzerDriverHelper.IsAnalyzerExceptionDiagnostic(args.Diagnostic));
+            
+            var diagnosticData = DiagnosticData.Create(args.Workspace, args.Diagnostic);
+            ImmutableArray<DiagnosticData> existingDiagnostics;
+            if (_analyzerHostDiagnosticsMap.TryGetValue(args.FaultedAnalyzer, out existingDiagnostics))
+            {
+                if (existingDiagnostics.Contains(diagnosticData))
+                {
+                    // don't fire duplicate diagnostics.
+                    return;
+                }
+            }
+            else
+            {
+                existingDiagnostics = ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            var dxs = ImmutableInterlocked.AddOrUpdate(ref _analyzerHostDiagnosticsMap,
+                args.FaultedAnalyzer,
+                ImmutableArray.Create(diagnosticData),
+                (a, existing) =>
+                {
+                    var newDiags = existing.Add(diagnosticData).Distinct();
```

Comment:
Ah, I wanted to switch this to use an immutable set representation, but left it at array. Let me see if I can switch the data structure and simplify this method.